{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "762878ea-de18-4d42-be65-eea64bbe6338",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "import shutil\n",
    "import pickle\n",
    "import random\n",
    "from pathlib import Path\n",
    "from collections import defaultdict\n",
    "from itertools import combinations, combinations_with_replacement, product, chain\n",
    "import urllib\n",
    "\n",
    "# Visualization and data analysis\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Patch\n",
    "import seaborn as sns\n",
    "from tqdm import tqdm\n",
    "from concurrent.futures import ProcessPoolExecutor\n",
    "\n",
    "# Proteomics libraries\n",
    "import pyopenms as pms\n",
    "from pyopenms import MSExperiment, MzMLFile\n",
    "from pyteomics import mzml\n",
    "from pyteomics.mass.mass import std_aa_mass\n",
    "import spectrum_utils as su"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c7b3385c-4f63-41f5-9abb-d2cb914c1551",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install seaborn"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b8d6d7a-be0f-448d-9009-75a13e82ce9e",
   "metadata": {},
   "source": [
    "# A deep dive into limitations of modification searching for DIA data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d41d0ba-c5ed-42a9-8733-ffc8dce7d960",
   "metadata": {},
   "source": [
    "Data-Dependent Acquisition (DDA) and Data-Independent Acquisition (DIA) are the two most common acqisition methods in mass spectrometry. Especially in recent years DIA has become more popular for its ability to be highly reproducible and potentially be more sensitive. The data from DIA is however much more complex to interpret and needs sophisticated data analysis methods to reach this potential.\n",
    "\n",
    "DDA vs. DIA: Fundamental Differences\n",
    "\n",
    "Data-Dependent Acquisition (DDA) generally selects ions based on their intensity from an MS1 scan. Based on the m/z ions are isolated within a specified isolation windows, these precursor ions are then fragmented (MS2), and the resulting spectra are used to identify peptides. This method focuses on the most abundant ions in each scan cycle. This is shown in the first figure where the DDA acquisition cycle (bottom) alternates between a full scan (MS1) and targeted MS2 scans of selected ions. It is important to note that DDA still has isolation windows wide enough to isolate multiple precursors of different peptidoform origin.\n",
    "\n",
    "In contrast, Data-Independent Acquisition (DIA) collects data across a wider range of m/z values without prior ion selection. Instead, DIA is preconfigured to specific isolation windows, fragmenting all ions in these preconfigured isolation windows. DIA allows for more comprehensive and unbiased acquisition, capturing information from all ions, which increases sensitivity and data completeness compared to DDA. However, the complexity of the data makes it harder to assign spectra to specific peptides due to overlapping signals and unknown origins of signals.\n",
    "\n",
    "![dda_vs_dia](img/1_ddavsdia.png)\n",
    "\n",
    "Open Modification Searches: Challenges\n",
    "\n",
    "In both DDA and DIA, the identification of modified peptides is challenging. These challenges become even more pronounced as the number of modifications included in the search space increases, with ambiguity growing disproportionately for each additional modification. Furthermore, when searching for a large number of modifications there are not only issues with identification ambiguity, but also the search space introduces computational challenges.\n",
    "\n",
    "![issues search](img/2_issues_search.png)\n",
    "\n",
    "In DDA, the identification is typically more straightforward since it aims to isolate a single peptidoform for fragmentation. This selective nature helps narrow down the search space. However, in DIA, all ions in a given m/z window are fragmented simultaneously, leading to more complex spectra. The broader acquisition means more overlapping peaks and increased ambiguity in the interpretation of modification sites, as the data includes multiple co-eluting peptides.\n",
    "\n",
    "Computational Issues and Ambiguities\n",
    "\n",
    "The third figure highlights one significant computational issue: ambiguity in peptide identification during open modification searches. For example, a peptide sequence, such as EEIFGPVM[Oxidation]QLIK and EEIFGP[Oxidation]VQPLFK. Both sequences are equally valid, but resolving them requires careful consideration of retention time and ion fragmentation patterns. Which can be (partly) resolved with predictions, but as we will see later this is still very difficult for DIA data.\n",
    "\n",
    "![issues search](img/3_resolving_ambiguity.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "65f5675b-27b4-4ae6-9fd4-8bac223ed27a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing: LFQ_Orbitrap_AIF_Human_01.mzML.gz\n",
      "Extracted: C:\\Users\\robbi\\OneDrive - UGent\\Documents\\2025\\Conferences\\eupa\\analysis\\EUPA_2025_DIA\\workshop\\LFQ_Orbitrap_AIF_Human_01.mzML\n",
      "\n",
      "Processing: LFQ_Orbitrap_DDA_Human_01.mzML.gz\n",
      "Extracted: C:\\Users\\robbi\\OneDrive - UGent\\Documents\\2025\\Conferences\\eupa\\analysis\\EUPA_2025_DIA\\workshop\\LFQ_Orbitrap_DDA_Human_01.mzML\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# List of URLs to download and extract\n",
    "urls = [\n",
    "    \"https://genesis.ugent.be/uvpublicdata/workshop_2025_eupa/LFQ_Orbitrap_AIF_Human_01.mzML.gz\",\n",
    "    \"https://genesis.ugent.be/uvpublicdata/workshop_2025_eupa/LFQ_Orbitrap_DDA_Human_01.mzML.gz\"\n",
    "]\n",
    "\n",
    "for url in urls:\n",
    "    # Determine filenames\n",
    "    output_gz = Path(url.split(\"/\")[-1])\n",
    "    output_mzml = output_gz.with_suffix(\"\")  # Remove '.gz'\n",
    "\n",
    "    print(f\"Processing: {output_gz.name}\")\n",
    "\n",
    "    # Step 1: Download the file\n",
    "    with urllib.request.urlopen(url) as response, open(output_gz, 'wb') as out_file:\n",
    "        shutil.copyfileobj(response, out_file)\n",
    "\n",
    "    # Step 2: Decompress the file\n",
    "    with gzip.open(output_gz, 'rb') as f_in, open(output_mzml, 'wb') as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)\n",
    "\n",
    "    # Optional: remove the .gz file\n",
    "    output_gz.unlink()\n",
    "\n",
    "    print(f\"Extracted: {output_mzml.resolve()}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0d91d69-ef26-42ee-aa94-8832564d1b26",
   "metadata": {},
   "source": [
    "# Visualize MS1 & MS2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9da55c90-643a-40df-8b9f-3b40864445d1",
   "metadata": {},
   "source": [
    "For this workshop we will use samples from human K562 cells. The data were acquired on a Thermo Orbitrap QE HF-X instrument in both DDA and DIA (AIF) mode with 8 m/z staggered windows. More details about the dataset can be found in [Van Puyvelde et al.](https://www.nature.com/articles/s41597-022-01216-6).\n",
    "![data source](img/4_datasource_bvp.png)\n",
    "\n",
    "First we will have a look at the special nature of DIA data where the isolation window is set much wider and the isolation windows follow up each other in a iterative fashion. This is important as likely many more precursors are within the same isolation window and the isolation is consistent between experiments and cover the full m/z range of interest.\n",
    "\n",
    "Lets first read the .mzML file of the run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a8a2dc48-e34d-48b0-b8f6-bc8a7008f766",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to your mzML file\n",
    "mzml_file = 'LFQ_Orbitrap_AIF_Human_01.mzML'\n",
    "\n",
    "def read_mzmzl(mzml_file,break_after_n_spectra=50000):\n",
    "    ms1_to_scan = {}\n",
    "    ms1_to_ms2 = {}\n",
    "    \n",
    "    # Open the mzML file and read the spectra\n",
    "    with mzml.MzML(mzml_file) as reader:\n",
    "        for idx, spectrum in enumerate(reader):\n",
    "            if spectrum['ms level'] == 1:\n",
    "                ms_1_scan_nr = int(spectrum.get('id', '').split('scan=')[1].split()[0])\n",
    "                ms1_to_scan[ms_1_scan_nr] = spectrum\n",
    "            elif spectrum['ms level'] == 2:\n",
    "                ms_2_scan_nr = int(spectrum.get('id', '').split('scan=')[1].split()[0])\n",
    "                if ms_1_scan_nr in ms1_to_ms2.keys():\n",
    "                    ms1_to_ms2[ms_1_scan_nr][ms_2_scan_nr] = spectrum\n",
    "                else:\n",
    "                    ms1_to_ms2[ms_1_scan_nr] = {}\n",
    "                    ms1_to_ms2[ms_1_scan_nr][ms_2_scan_nr] = spectrum\n",
    "    \n",
    "            # Only process first n spectra for performance\n",
    "            if idx > break_after_n_spectra:\n",
    "                break\n",
    "    return ms1_to_scan, ms1_to_ms2\n",
    "\n",
    "ms1_to_scan, ms1_to_ms2 = read_mzmzl(mzml_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff00231f-13c9-4ecc-9d5f-c893d3187393",
   "metadata": {},
   "source": [
    "Next we define some helper functions for plotting. There are in total three functions:\n",
    "\n",
    "1. **get_random_ms_specs** - select a random MS1 and an MS2 (that are \"linked\") from the read file. Return all relevant information for plotting\n",
    "2. **get_ms_specs_scan** - based on a specific MS1 and MS2 scan number return the spectra\n",
    "3. **create_ms_plot** - plot the MS1 and MS2 spectra. For the MS1 indicate the region that is isolated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8694a499-5245-4308-bd15-1f4ad52229e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_random_ms_specs(ms1_to_scan,ms1_to_ms2):\n",
    "    # Select a random MS1 spectrum\n",
    "    random_ms1 = random.choice(list(ms1_to_scan.keys()))\n",
    "    ms1_spectrum = ms1_to_scan[random_ms1]\n",
    "    mz1, intensity1 = np.array(ms1_spectrum['m/z array']), np.array(ms1_spectrum['intensity array'])\n",
    "    \n",
    "    # Assuming the precursor information is available in the MS1 spectrum\n",
    "    precursor_mz1 = ms1_spectrum.get('precursorMz', np.nan)  # Precursor m/z (if available)\n",
    "    precursor_charge1 = ms1_spectrum.get('precursorCharge', 1)  # Precursor charge (if available)\n",
    "    \n",
    "    # Select a random MS2 spectrum related to the random MS1 spectrum\n",
    "    random_ms2 = random.choice(list(ms1_to_ms2[random_ms1].keys()))\n",
    "    ms2_spectrum = ms1_to_ms2[random_ms1][random_ms2]\n",
    "    mz2, intensity2 = np.array(ms2_spectrum['m/z array']), np.array(ms2_spectrum['intensity array'])\n",
    "    \n",
    "    # Get the precursor m/z of the MS2 spectrum (for the isolation window)\n",
    "    precursor_mz2 = ms2_spectrum['precursorList']['precursor'][0]['isolationWindow']['isolation window target m/z']\n",
    "    isolation_width = ms2_spectrum['precursorList']['precursor'][0]['isolationWindow']['isolation window lower offset']\n",
    "\n",
    "    return mz1, intensity1, ms1_spectrum, random_ms1, mz2, intensity2, ms2_spectrum, random_ms2, precursor_mz2, isolation_width\n",
    "\n",
    "def get_ms_specs_scan(ms1_to_scan,ms1_to_ms2,ms1_scan,ms2_scan):\n",
    "    ms1_spectrum = ms1_to_scan[ms1_scan]\n",
    "    mz1, intensity1 = np.array(ms1_spectrum['m/z array']), np.array(ms1_spectrum['intensity array'])\n",
    "    \n",
    "    # Assuming the precursor information is available in the MS1 spectrum\n",
    "    precursor_mz1 = ms1_spectrum.get('precursorMz', np.nan)  # Precursor m/z (if available)\n",
    "    precursor_charge1 = ms1_spectrum.get('precursorCharge', 1)  # Precursor charge (if available)\n",
    "    \n",
    "    ms2_spectrum = ms1_to_ms2[ms1_scan][ms2_scan]\n",
    "    mz2, intensity2 = np.array(ms2_spectrum['m/z array']), np.array(ms2_spectrum['intensity array'])\n",
    "    \n",
    "    # Get the precursor m/z of the MS2 spectrum (for the isolation window)\n",
    "    precursor_mz2 = ms2_spectrum['precursorList']['precursor'][0]['isolationWindow']['isolation window target m/z']\n",
    "    isolation_width = ms2_spectrum['precursorList']['precursor'][0]['isolationWindow']['isolation window lower offset']\n",
    "\n",
    "    return mz1, intensity1, ms1_spectrum, mz2, intensity2, ms2_spectrum, precursor_mz2, isolation_width\n",
    "\n",
    "def create_ms_plot(mz1,intensity1, ms1_spectrum, mz2, intensity2, ms2_spectrum, precursor_mz2, isolation_width):\n",
    "    # Plot MS1 spectrum using stem plot (vertical lines for each peak)\n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15/2, 6/2))\n",
    "    \n",
    "    markerline, stemlines, baseline = ax1.stem(mz1, intensity1, basefmt=\" \", markerfmt=\" \", linefmt='C0-')\n",
    "    plt.setp(stemlines, 'linewidth', 0.5)\n",
    "    # Highlight the isolation window around the precursor in MS1\n",
    "    \n",
    "    # Highlight the isolation window around the precursor in MS2 (for MS1 plot)\n",
    "    if not np.isnan(precursor_mz2):\n",
    "        ax1.axvspan(precursor_mz2 - isolation_width, precursor_mz2 + isolation_width, color='red', alpha=0.5, label='Precursor Isolation Window')\n",
    "    \n",
    "    ax1.set_xlabel('m/z')\n",
    "    ax1.set_ylabel('Intensity')\n",
    "    ax1.set_title(f'MS1 Spectrum: Scan {random_ms1}')\n",
    "    ax1.grid(True)\n",
    "    ax1.legend()\n",
    "    \n",
    "    markerline, stemlines, baseline = ax2.stem(mz2, intensity2, basefmt=\" \", markerfmt=\" \", linefmt='C1-')\n",
    "    plt.setp(stemlines, 'linewidth', 0.5)\n",
    "    ax2.set_xlabel('m/z')\n",
    "    ax2.set_ylabel('Intensity')\n",
    "    ax2.set_title(f'MS2 Spectrum: Scan {random_ms2}')\n",
    "    ax2.grid(True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5fbc52ee-e5a2-4cfa-929a-5a59d7f2dbb5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ms1_to_scan' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[8]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m mz1, intensity1, ms1_spectrum, random_ms1, mz2, intensity2, ms2_spectrum, random_ms2, precursor_mz2, isolation_width = get_random_ms_specs(\u001b[43mms1_to_scan\u001b[49m,ms1_to_ms2)\n\u001b[32m      2\u001b[39m create_ms_plot(mz1,intensity1, ms1_spectrum, mz2, intensity2, ms2_spectrum, precursor_mz2, isolation_width)\n\u001b[32m      4\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[32m0\u001b[39m,\u001b[32m50\u001b[39m):\n",
      "\u001b[31mNameError\u001b[39m: name 'ms1_to_scan' is not defined"
     ]
    }
   ],
   "source": [
    "mz1, intensity1, ms1_spectrum, random_ms1, mz2, intensity2, ms2_spectrum, random_ms2, precursor_mz2, isolation_width = get_random_ms_specs(ms1_to_scan,ms1_to_ms2)\n",
    "create_ms_plot(mz1,intensity1, ms1_spectrum, mz2, intensity2, ms2_spectrum, precursor_mz2, isolation_width)\n",
    "\n",
    "for i in range(0,50):\n",
    "    try:\n",
    "        mz1, intensity1, ms1_spectrum, mz2, intensity2, ms2_spectrum, precursor_mz2, isolation_width = get_ms_specs_scan(ms1_to_scan,ms1_to_ms2,random_ms1,random_ms1+i)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    create_ms_plot(mz1,intensity1, ms1_spectrum, mz2, intensity2, ms2_spectrum, precursor_mz2, isolation_width)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "410741d4-8f4f-4849-a895-a1b367bcd38a",
   "metadata": {},
   "source": [
    "# Count peaks in MS1 and MS2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fccaa4e-4ff4-4796-86ea-d8fd81b483de",
   "metadata": {},
   "source": [
    "![count](https://media1.tenor.com/m/CSYSqLuo6fwAAAAC/sesame-street-muppets.gif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc505a74-97b3-47cd-afd7-b9da62936918",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_file(mzml_path):\n",
    "    exp = MSExperiment()\n",
    "    MzMLFile().load(mzml_path, exp)\n",
    "    num_peaks = []\n",
    "\n",
    "    for s in tqdm(exp.getSpectra()):\n",
    "        if s.getMSLevel() == 1:\n",
    "            ms1_spectrum = s.get_peaks()[0]\n",
    "            continue\n",
    "\n",
    "        lower_val = (\n",
    "            s.getPrecursors()[0].getMZ()\n",
    "            - s.getPrecursors()[0].getIsolationWindowLowerOffset()\n",
    "        )\n",
    "        upper_val = (\n",
    "            s.getPrecursors()[0].getMZ()\n",
    "            + s.getPrecursors()[0].getIsolationWindowUpperOffset()\n",
    "        )\n",
    "\n",
    "        try:\n",
    "            num_peaks.append(\n",
    "                sum([1 for mz in ms1_spectrum if lower_val < mz < upper_val])\n",
    "            )\n",
    "        except:\n",
    "            continue\n",
    "\n",
    "    return num_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7a18627-cd37-4570-abe2-85a707375635",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_file(mzml_path, ms1_mode=True):\n",
    "    num_peaks = []\n",
    "    ms1_spectrum = None\n",
    "\n",
    "    # Open mzML file using pyteomics\n",
    "    with mzml.MzML(mzml_path) as reader:\n",
    "        for spectrum in tqdm(reader):\n",
    "            # Process MS1 spectra\n",
    "            if spectrum['ms level'] == 1:\n",
    "                ms1_spectrum = spectrum['m/z array']\n",
    "                continue\n",
    "\n",
    "            # Process MS2 spectra\n",
    "            if spectrum['ms level'] == 2 and ms1_spectrum is not None and ms1_mode:\n",
    "                precursor_mz = spectrum['precursorList']['precursor'][0]['isolationWindow']['isolation window target m/z']\n",
    "                offset = spectrum['precursorList']['precursor'][0]['isolationWindow']['isolation window lower offset']\n",
    "\n",
    "                lower_val = precursor_mz - offset\n",
    "                upper_val = precursor_mz + offset\n",
    "\n",
    "                # Count the number of peaks from MS1 spectrum within the isolation window\n",
    "                num_peaks.append(\n",
    "                    sum([1 for mz in ms1_spectrum if lower_val < mz < upper_val])\n",
    "                )\n",
    "            if spectrum['ms level'] == 2 and not ms1_mode:\n",
    "                num_peaks.append(\n",
    "                    sum([1 for mz in spectrum['m/z array']])\n",
    "                )\n",
    "                \n",
    "    return num_peaks\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa7bf6b8-de45-4f64-968e-2db59354ea7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mzml_file_dia = \"./LFQ_Orbitrap_AIF_Human_01.mzML\"\n",
    "mzml_file_dda = \"./LFQ_Orbitrap_DDA_Human_01.mzML\"\n",
    "\n",
    "num_peaks_dia_ms1 = process_file(mzml_file_dia)\n",
    "num_peaks_dda_ms1 = process_file(mzml_file_dda)\n",
    "\n",
    "num_peaks_dia_ms2 = process_file(mzml_file_dia, ms1_mode=False)\n",
    "num_peaks_dda_ms2 = process_file(mzml_file_dda, ms1_mode=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b845d16-89ed-433f-acbb-56cc5685a153",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_peaks_dict = {}\n",
    "\n",
    "num_peaks_dict[\"LFQ_Orbitrap_AIF_Human_01.mzML_ms1\"] = num_peaks_dia_ms1\n",
    "num_peaks_dict[\"LFQ_Orbitrap_DDA_Human_01.mzML_ms1\"] = num_peaks_dda_ms1\n",
    "\n",
    "num_peaks_dict[\"LFQ_Orbitrap_AIF_Human_01.mzML_ms2\"] = num_peaks_dia_ms2\n",
    "num_peaks_dict[\"LFQ_Orbitrap_DDA_Human_01.mzML_ms2\"] = num_peaks_dda_ms2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1c55ce5-0955-40f4-b2b8-dfc5940450a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "dia_key = \"LFQ_Astral_DIA_15min_50ng_Condition_A_REP1.mzML_ms1\"\n",
    "dda_key = \"LFQ_Astral_DDA_15min_50ng_Condition_A_REP1.mzML_ms1\"\n",
    "\n",
    "plt.hist(\n",
    "    num_peaks_dict[dda_key],\n",
    "    bins=list(range(0, 260)),\n",
    "    color=\"blue\",\n",
    "    alpha=0.5,\n",
    "    label=\"DDA Orbitrap\",\n",
    ")\n",
    "plt.hist(\n",
    "    num_peaks_dict[dia_key],\n",
    "    bins=list(range(0, 260)),\n",
    "    color=\"red\",\n",
    "    alpha=0.5,\n",
    "    label=\"DIA Orbitrap\",\n",
    ")\n",
    "\n",
    "plt.title(\"MS1 peaks in a spectrum\")\n",
    "\n",
    "plt.legend()\n",
    "plt.xlabel(\"Peaks in spectrum (#)\")\n",
    "plt.ylabel(\"Frequency (#)\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"img/ms1_peaks_lfq_orbitrap.svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7265545-6e9e-479f-ba57-f243d43895fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "dia_key = \"LFQ_Astral_DIA_15min_50ng_Condition_A_REP1.mzML_ms2\"\n",
    "dda_key = \"LFQ_Astral_DDA_15min_50ng_Condition_A_REP1.mzML_ms2\"\n",
    "\n",
    "plt.hist(\n",
    "    num_peaks_dict[dda_key],\n",
    "    bins=list(range(0, 260)),\n",
    "    color=\"blue\",\n",
    "    alpha=0.5,\n",
    "    label=\"DDA Orbitrap\",\n",
    ")\n",
    "plt.hist(\n",
    "    num_peaks_dict[dia_key],\n",
    "    bins=list(range(0, 260)),\n",
    "    color=\"red\",\n",
    "    alpha=0.5,\n",
    "    label=\"DIA Orbitrap\",\n",
    ")\n",
    "\n",
    "plt.title(\"MS1 peaks in a spectrum\")\n",
    "\n",
    "plt.legend()\n",
    "plt.xlabel(\"Peaks in spectrum (#)\")\n",
    "plt.ylabel(\"Frequency (#)\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"img/ms1_peaks_lfq_orbitrap.svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccc85166-10d5-4d10-bdc7-7930bd32f054",
   "metadata": {},
   "source": [
    "# Ambiguity - multi interpretation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e867be7b-1a80-4235-8d48-54e7e2f165fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acid_masses = {\n",
    "    # Standard amino acids\n",
    "    \"A\": 71.03711,\n",
    "    \"R\": 156.10111,\n",
    "    \"N\": 114.04293,\n",
    "    \"D\": 115.02694,\n",
    "    \"C\": 103.00919,\n",
    "    \"E\": 129.04259,\n",
    "    \"Q\": 128.05858,\n",
    "    \"G\": 57.02146,\n",
    "    \"H\": 137.05891,\n",
    "    \"I\": 113.08406,\n",
    "    \"L\": 113.08406,\n",
    "    \"K\": 128.09496,\n",
    "    \"M\": 131.04049,\n",
    "    \"F\": 147.06841,\n",
    "    \"P\": 97.05276,\n",
    "    \"S\": 87.03203,\n",
    "    \"T\": 101.04768,\n",
    "    \"W\": 186.07931,\n",
    "    \"Y\": 163.06333,\n",
    "    \"V\": 99.06841,\n",
    "\n",
    "    # Modified residues (accurate total masses)\n",
    "    \"C[Carbamidomethyl]\": 160.03065,     # 103.00919 + 57.02146\n",
    "    \"M[Oxidation]\": 147.03540,           # 131.04049 + 15.99491\n",
    "    \"H[Oxidation]\": 153.05382,           # 137.05891 + 15.99491\n",
    "    \"Q[Pyro-glu]\": 111.03203,            # 128.05858 − 17.02655\n",
    "    \"E[Pyro-glu]\": 111.03203,            # 129.04259 − 18.01056\n",
    "    \"W[Oxidation]\": 202.07422,           # 186.07931 + 15.99491\n",
    "    \"N[Deamidated]\": 115.02695,          # 114.04293 + 0.98402\n",
    "    \"Q[Deamidated]\": 129.04260,          # 128.05858 + 0.98402\n",
    "    \"S[Phospho]\": 166.99836,             # 87.03203 + 79.96633\n",
    "    \"T[Phospho]\": 181.01398,             # 101.04768 + 79.96633\n",
    "    \"Y[Phospho]\": 243.02966,             # 163.06333 + 79.96633\n",
    "    \"K[Acetyl]\": 170.10552,              # 128.09496 + 42.01056\n",
    "    \"K[Methyl]\": 142.11061,              # 128.09496 + 14.01565\n",
    "    \"K[Dimethyl]\": 156.12626,            # 128.09496 + 28.03130\n",
    "    \"K[Trimethyl]\": 170.14191,           # 128.09496 + 42.04695\n",
    "    \"K[Propionyl]\": 185.11642,           # 128.09496 + 57.02146\n",
    "    \"K[Butyryl]\": 199.13031,             # 128.09496 + 71.03535\n",
    "    \"K[Formyl]\": 156.10572,              # 128.09496 + 28.01076\n",
    "    \"K[Malonyl]\": 186.11146,             # 128.09496 + 58.01650\n",
    "    \"K[Succinyl]\": 200.12535,            # 128.09496 + 72.03039\n",
    "    \"R[Methyl]\": 170.11676,              # 156.10111 + 14.01565\n",
    "    \"R[Dimethyl]\": 184.13241,            # 156.10111 + 28.03130\n",
    "    \"R[Trimethyl]\": 198.14806,           # 156.10111 + 42.04695\n",
    "}\n",
    "\n",
    "all_masses = {**amino_acid_masses}\n",
    "\n",
    "# Converting to a sorted list of masses for easier indexing\n",
    "sorted_masses = sorted(all_masses.items(), key=lambda x: x[1])\n",
    "amino_acid_deltas = [mass for name, mass in sorted_masses]\n",
    "amino_acid_deltas_names = [name for name, mass in sorted_masses]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dbc698a-4635-4e5a-bc9d-0d79866baa34",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_pairwise_distances(mzs, amino_acid_deltas, tolerance=0.005, max_mz=300.0):\n",
    "    mzs = np.array(mzs)\n",
    "    if len(mzs) < 2:  # Ensure there are at least two m/z values to compare\n",
    "        return np.zeros(len(amino_acid_deltas), dtype=int)\n",
    "    mz_diffs = np.abs(mzs[:, np.newaxis] - mzs)\n",
    "    mz_diffs = np.triu(mz_diffs, k=1)\n",
    "    mz_i, mz_j = np.where((mz_diffs > 0) & (mz_diffs <= max_mz))\n",
    "    if not mz_i.size:  # No valid differences found\n",
    "        return np.zeros(len(amino_acid_deltas), dtype=int)\n",
    "    mz_diffs = mz_diffs[mz_i, mz_j]\n",
    "    comparison = np.abs(mz_diffs[:, np.newaxis] - amino_acid_deltas) <= tolerance\n",
    "    return np.sum(comparison, axis=0)\n",
    "\n",
    "def bin_and_sum_intensities(pairs):\n",
    "    bins = np.arange(0, 300.005, 0.005)\n",
    "    intensity_sums = np.zeros(len(bins))\n",
    "    for mz1, mz2, diff, intensity_sum in pairs:\n",
    "        idx = int(diff / 0.005)\n",
    "        intensity_sums[idx] += intensity_sum\n",
    "    return intensity_sums\n",
    "\n",
    "def process_file(mzml_path, amino_acid_deltas):\n",
    "    exp = MSExperiment()\n",
    "    MzMLFile().load(mzml_path, exp)\n",
    "    spectra = [\n",
    "        (s.get_peaks()[0], s.get_peaks()[1])  # select_top_peaks\n",
    "        for s in exp.getSpectra()\n",
    "        if s.getMSLevel() == 2\n",
    "    ]\n",
    "    \n",
    "    results_matrix = []\n",
    "    for mzs, intensities in tqdm(spectra):\n",
    "        if len(mzs) > 1:\n",
    "            results_matrix.append(calculate_pairwise_distances(mzs, amino_acid_deltas))\n",
    "        else:\n",
    "            results_matrix.append(\n",
    "                np.zeros(len(amino_acid_deltas), dtype=int)\n",
    "            )  # Ensure consistent shape\n",
    "    return results_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c5ea927-739d-4c57-abb3-588c4edbb95c",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_dict = {}\n",
    "files = [\n",
    "    \"./LFQ_Orbitrap_AIF_Human_01.mzML\",\n",
    "    \"./LFQ_Orbitrap_DDA_Human_01.mzML\"\n",
    "]\n",
    "result_dir = \"./\"\n",
    "\n",
    "for file in files:\n",
    "    result = process_file(file, amino_acid_deltas)\n",
    "    result = pd.DataFrame(result)\n",
    "    result.columns = amino_acid_deltas_names\n",
    "    results_dict[os.path.basename(file)] = result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8255e3e7-4088-46a8-900f-f99c08a27c93",
   "metadata": {},
   "outputs": [],
   "source": [
    "dia_key = \"LFQ_Orbitrap_AIF_Human_01.mzML\"\n",
    "dda_key = \"LFQ_Orbitrap_DDA_Human_01.mzML\"\n",
    "\n",
    "# Initialize the figure and axis for the plot\n",
    "plt.figure(figsize=(10, 5))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Lists to store values for plotting the distributions\n",
    "dia_values = []\n",
    "dda_values = []\n",
    "labels = []\n",
    "\n",
    "for i in range(20, len(amino_acid_masses.keys()) + 1):\n",
    "    selected_cols = list(amino_acid_masses.keys())[0:i]\n",
    "\n",
    "    # Calculate the sums of the selected columns\n",
    "    dia_count_all = results_dict[dia_key][selected_cols].sum(axis=1)\n",
    "    dda_count_all = results_dict[dda_key][selected_cols].sum(axis=1)\n",
    "\n",
    "    # Filter sums greater than 10 and collect values\n",
    "    dia_values.append(dia_count_all[dia_count_all > 10].values)\n",
    "    dda_values.append(dda_count_all[dda_count_all > 10].values)\n",
    "\n",
    "    labels.append(str(i))\n",
    "\n",
    "# Plotting boxplots\n",
    "positions_dia = [x - 0.2 for x in range(20, len(amino_acid_masses.keys()) + 1)]\n",
    "positions_dda = [x + 0.2 for x in range(20, len(amino_acid_masses.keys()) + 1)]\n",
    "central_positions = [x for x in range(20, len(amino_acid_masses.keys()) + 1)]\n",
    "\n",
    "ax.boxplot(\n",
    "    dia_values,\n",
    "    positions=positions_dia,\n",
    "    widths=0.4,\n",
    "    patch_artist=True,\n",
    "    boxprops=dict(facecolor=\"red\"),\n",
    "    medianprops=dict(color=\"black\"),\n",
    "    showfliers=False,\n",
    ")\n",
    "ax.boxplot(\n",
    "    dda_values,\n",
    "    positions=positions_dda,\n",
    "    widths=0.4,\n",
    "    patch_artist=True,\n",
    "    boxprops=dict(facecolor=\"blue\"),\n",
    "    medianprops=dict(color=\"black\"),\n",
    "    showfliers=False,\n",
    ")\n",
    "\n",
    "# Custom legend\n",
    "legend_elements = [\n",
    "    Patch(facecolor=\"red\", edgecolor=\"black\", label=\"DIA\"),\n",
    "    Patch(facecolor=\"blue\", edgecolor=\"black\", label=\"DDA\"),\n",
    "]\n",
    "ax.legend(handles=legend_elements, loc=\"upper left\")\n",
    "\n",
    "# Setting the plot limits, labels, title, and x-ticks\n",
    "ax.set_ylim(0, 4000)\n",
    "ax.set_xlabel(\"Number of Selected Columns\")\n",
    "ax.set_ylabel(\"Summed Values\")\n",
    "ax.set_title(\"Empirical Differential of Summed Values Over Columns\")\n",
    "ax.set_xticks(central_positions)\n",
    "ax.set_xticklabels(labels, rotation=45)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "560c4af3-6dbe-48a4-8e11-e5c73544daae",
   "metadata": {},
   "source": [
    "# Search space size "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2c6a819-ed72-45f2-84b5-61b600a5cb63",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6765f25-e875-4b5d-87a0-4846fb9d4ebf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tryptic_digest_pyopenms(\n",
    "    file_path,\n",
    "    min_len=5,\n",
    "    max_len=50,\n",
    "    missed_cleavages=2,\n",
    "    decoy_method=\"reverse\",\n",
    "    decoy_prefix=\"rev_\",\n",
    "    seq_types=[\"original\", \"decoy\"],\n",
    "):\n",
    "    # Read the FASTA file\n",
    "    fasta = pms.FASTAFile()\n",
    "    entries = []\n",
    "    fasta.load(file_path, entries)\n",
    "\n",
    "    # Set up the enzyme digestion\n",
    "    digestor = pms.ProteaseDigestion()\n",
    "    digestor.setEnzyme(\"Trypsin\")\n",
    "    digestor.setMissedCleavages(missed_cleavages)\n",
    "\n",
    "    peptides = []\n",
    "    for entry in entries:\n",
    "        # Process both original and decoy sequences\n",
    "        for seq_type in seq_types:\n",
    "            if seq_type == \"original\":\n",
    "                protein_sequence = str(entry.sequence)\n",
    "            else:\n",
    "                if decoy_method == \"reverse\":\n",
    "                    protein_sequence = str(entry.sequence)[::-1]\n",
    "                elif decoy_method == \"scramble\":\n",
    "                    seq_list = list(str(entry.sequence))\n",
    "                    random.shuffle(seq_list)\n",
    "                    protein_sequence = \"\".join(seq_list)\n",
    "                else:\n",
    "                    raise ValueError(\n",
    "                        \"Invalid decoy method. Choose 'reverse' or 'scramble'.\"\n",
    "                    )\n",
    "\n",
    "            protein_name = entry.identifier.split()[\n",
    "                0\n",
    "            ]  # Adjust based on your FASTA format\n",
    "\n",
    "            # Perform the tryptic digest\n",
    "            result = []\n",
    "            digestor.digest(pms.AASequence.fromString(protein_sequence), result)\n",
    "\n",
    "            for peptide in result:\n",
    "                peptide_sequence = str(peptide.toString())\n",
    "                len_pep_seq = len(peptide_sequence)\n",
    "                start = protein_sequence.find(peptide_sequence)\n",
    "                end = start + len_pep_seq\n",
    "                if \"X\" in peptide_sequence:\n",
    "                    continue\n",
    "                if len_pep_seq >= min_len and len_pep_seq <= max_len:\n",
    "                    if seq_type == \"original\":\n",
    "                        peptides.append(\n",
    "                            (\n",
    "                                protein_name,\n",
    "                                start,\n",
    "                                end,\n",
    "                                f\"{protein_name}|{start}|{end}\",\n",
    "                                peptide_sequence,\n",
    "                            )\n",
    "                        )\n",
    "                    else:\n",
    "                        peptides.append(\n",
    "                            (\n",
    "                                f\"{decoy_prefix}{protein_name}\",\n",
    "                                start,\n",
    "                                end,\n",
    "                                f\"{decoy_prefix}{protein_name}|{start}|{end}\",\n",
    "                                peptide_sequence,\n",
    "                            )\n",
    "                        )\n",
    "    return peptides\n",
    "    \n",
    "def calculate_possibilities(possibilities,max_mods=3):\n",
    "    N = len(possibilities)  # Total number of positions\n",
    "    total = 0\n",
    "\n",
    "    # Iterate over the number of positions that can vary (0 to 3)\n",
    "    for k in range(0, max_mods+1):\n",
    "        # Generate all combinations of positions that can vary\n",
    "        for varying_positions in combinations(range(N), k):\n",
    "            # Calculate the product of possibilities for the chosen positions\n",
    "            product = 1\n",
    "            for pos in varying_positions:\n",
    "                product *= possibilities[pos]\n",
    "            # Add this product to the total sum\n",
    "            total += product\n",
    "\n",
    "    return total\n",
    "\n",
    "def modification_list_to_dict(all_amino_acids, modifications_search_space):\n",
    "    modification_dict = defaultdict(list)\n",
    "    for modification, amino_acids in modifications_search_space:\n",
    "        for aa in amino_acids:\n",
    "            modification_dict[aa].append(modification)\n",
    "\n",
    "    for aa in all_amino_acids:\n",
    "        modification_dict[aa].insert(0, \"\")\n",
    "\n",
    "    modification_dict = dict(modification_dict)\n",
    "    return modification_dict\n",
    "\n",
    "def apply_modifications(\n",
    "    peptides, modifications, charges=[2, 3, 4, 5], max_var_mods=3, max_len=30, min_len=7\n",
    "):\n",
    "    unique_peptides = set([p[-1] for p in peptides])\n",
    "    tot_num = 0\n",
    "    tot_num_fragments = 0\n",
    "\n",
    "    for idx, peptide_seq in enumerate(unique_peptides):\n",
    "        if len(peptide_seq) > max_len or len(peptide_seq) < min_len:\n",
    "            continue\n",
    "        all_pos = [len(modifications[aa]) - 1 for aa in peptide_seq]\n",
    "        poss = calculate_possibilities(all_pos, max_mods=max_var_mods) * len(charges)\n",
    "        tot_num += poss\n",
    "        tot_num_fragments += poss * len(peptide_seq) * 2\n",
    "\n",
    "    return tot_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82fc621d-ea5a-4aa9-be78-7a754aff281b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your modification data\n",
    "modifications_search_space = [\n",
    "    (\"[Oxidation]\", [\"M\"]),  # , \"P\"\n",
    "    (\"[OxidationP]\", [\"P\"]),  # , \"P\"\n",
    "    (\"[Carbamidomethyl]\", [\"C\"]),\n",
    "    (\"[Alkyl]\", [\"C\"]),\n",
    "    (\"[DeamidationN]\", [\"N\"]),\n",
    "    (\"[DeamidationQ]\", [\"Q\"]),\n",
    "    (\"[PhosphorylationS]\", [\"S\"]),\n",
    "    (\"[PhosphorylationT]\", [\"T\"]),\n",
    "    (\"[PhosphorylationY]\", [\"Y\"]),\n",
    "    (\"[Nitrosyl]\", [\"C\"]),\n",
    "    (\"[OxidationC]\", [\"C\"]),\n",
    "    (\"[Methyl]\", [\"K\"]),\n",
    "    (\"[MethylR]\", [\"R\"]),\n",
    "    (\"[Dimethyl]\", [\"K\"]),\n",
    "    (\"[DimethylR]\", [\"R\"]),\n",
    "    (\"[Trimethyl]\", [\"K\"]),\n",
    "    (\"[CitrullinationR]\", [\"R\"]),\n",
    "    (\"[Propionyl]\", [\"K\"]),\n",
    "    (\"[Butyryl]\", [\"K\"]),\n",
    "    (\"[Malonyl]\", [\"K\"]),\n",
    "    (\"[Succinyl]\", [\"K\"]),\n",
    "    (\"[Glutarylation]\", [\"K\"]),\n",
    "    (\"[Crotonyl]\", [\"K\"]),\n",
    "    (\"[Hydroxyisobutyryl]\", [\"K\"]),\n",
    "    (\"[Biotin]\", [\"K\"]),\n",
    "    (\"[GG]\", [\"K\"]),\n",
    "    (\"[NitroY]\", [\"Y\"]),\n",
    "]\n",
    "\n",
    "\n",
    "charge_min_range = range(1, 3)\n",
    "charge_max_range = [2]\n",
    "maximum_length_range = [35]\n",
    "max_var_mods_range = [1]\n",
    "missed_cleavages = [2]\n",
    "num_var_mods = range(0, len(modifications_search_space[0:5]))\n",
    "\n",
    "combinations_df = product(\n",
    "    charge_min_range,\n",
    "    charge_max_range,\n",
    "    maximum_length_range,\n",
    "    max_var_mods_range,\n",
    "    missed_cleavages,\n",
    "    num_var_mods,\n",
    ")\n",
    "\n",
    "df = pd.DataFrame(\n",
    "    combinations_df,\n",
    "    columns=[\n",
    "        \"charge_min\",\n",
    "        \"charge_max\",\n",
    "        \"maximum_length\",\n",
    "        \"max_var_mods\",\n",
    "        \"missed_cleavages\",\n",
    "        \"num_var_mods\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8c3e481-c129-4994-a9bd-08d504e59688",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_analyze = \"human_22032024.fasta\"\n",
    "\n",
    "# Define all standard amino acids\n",
    "all_amino_acids = set(\n",
    "    \"ABCDEFGHIKLMNPQRSTUVWXY\"\n",
    ")  # 20 standard amino acids + additional ones that can be in the fasta, they are ignored if they are not in there\n",
    "\n",
    "search_space_list = []\n",
    "for idx,r in tqdm(df.iterrows()):\n",
    "    i = r[\"num_var_mods\"]\n",
    "    # Read peptides from fasta using a tryptic digest function\n",
    "    peptides = tryptic_digest_pyopenms(\n",
    "        fasta_analyze,\n",
    "        min_len=6,\n",
    "        max_len=r[\"maximum_length\"],\n",
    "        missed_cleavages=int(r[\"missed_cleavages\"]),\n",
    "    )\n",
    "\n",
    "    modification_dict = modification_list_to_dict(\n",
    "        all_amino_acids, modifications_search_space[0:i+1]\n",
    "    )\n",
    "\n",
    "    # Apply modifications to peptides\n",
    "    all_peptidoforms = apply_modifications(\n",
    "        peptides,\n",
    "        modification_dict,\n",
    "        charges=list(range(r[\"charge_min\"], r[\"charge_max\"])),\n",
    "        max_var_mods=r[\"max_var_mods\"],\n",
    "    )\n",
    "    print(f\"Total number of unique peptidoforms: {all_peptidoforms}\")\n",
    "    search_space_list.append(all_peptidoforms)\n",
    "    \n",
    "df[\"search_space\"] = search_space_list\n",
    "df.to_csv(\"data/human_searchpace.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b12199c3-3999-46f5-afe1-0c330d4052ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_df_search_space = df[df[\"charge_min\"] == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c32b4270-f322-4fee-9430-d3767ab70b2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(sub_df_search_space[\"num_var_mods\"], sub_df_search_space[\"search_space\"],s=20,alpha=0.5)\n",
    "plt.plot(sub_df_search_space[\"num_var_mods\"], sub_df_search_space[\"search_space\"])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d89fa6c8-1529-47ca-91b9-8c84f3ec737a",
   "metadata": {},
   "source": [
    "# Prediction comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a5cb085-7924-4807-b9f1-e67e7a6043c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tr = pd.read_csv(\"prosit_ptm_2020.csv\")\n",
    "df_tr.fillna(\"\",inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68886d59-5002-4447-9aec-f6e6c7779d98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Separate unmodified and modified peptides\n",
    "df_unmod = df_tr[df_tr[\"modifications\"] == \"\"].rename(columns={\"tr\": \"tr_unmod\"})\n",
    "df_mod = df_tr[df_tr[\"modifications\"] != \"\"]\n",
    "\n",
    "# Merge on sequence to get corresponding unmodified tr for each modified one\n",
    "df_merged = pd.merge(df_mod, df_unmod[[\"seq\", \"tr_unmod\"]], on=\"seq\", how=\"left\")\n",
    "\n",
    "# Calculate RT difference\n",
    "df_merged[\"delta_tr\"] = df_merged[\"tr\"] - df_merged[\"tr_unmod\"]\n",
    "df_merged = df_merged.sort_values(by=\"seq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1819986e-b6c6-438d-a50a-02e5c20262fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_vp = df_exploded.dropna(subset=[\"delta_tr\", \"mod_types\"])\n",
    "\n",
    "# Limit to top modifications for clarity\n",
    "top_mods = df_vp[\"mod_types\"].value_counts().nlargest(10).index\n",
    "df_vp = df_vp[df_vp[\"mod_types\"].isin(top_mods)]\n",
    "\n",
    "# Function to identify outliers by IQR for each group\n",
    "def filter_outliers(group):\n",
    "    q1 = group[\"delta_tr\"].quantile(0.25)\n",
    "    q3 = group[\"delta_tr\"].quantile(0.75)\n",
    "    iqr = q3 - q1\n",
    "    lower_bound = q1 - 1.5 * iqr\n",
    "    upper_bound = q3 + 1.5 * iqr\n",
    "    # Return group without outliers\n",
    "    return group[(group[\"delta_tr\"] >= lower_bound) & (group[\"delta_tr\"] <= upper_bound)]\n",
    "\n",
    "# Filter outliers for violin plot only\n",
    "df_vp_filtered = df_vp.groupby(\"mod_types\").apply(filter_outliers).reset_index(drop=True)\n",
    "\n",
    "sns.set_style(\"whitegrid\")\n",
    "\n",
    "# Calculate median ΔRT per modification type\n",
    "mod_order = df_vp_filtered.groupby(\"mod_types\")[\"delta_tr\"].median().sort_values(ascending=False).index\n",
    "\n",
    "plt.figure(figsize=(9, 5.0))\n",
    "\n",
    "# Pass the sorted order to violinplot\n",
    "sns.violinplot(x=\"mod_types\", y=\"delta_tr\", data=df_vp_filtered, order=mod_order)\n",
    "\n",
    "# Optional: show all points with swarmplot (uncomment if desired)\n",
    "# sns.swarmplot(x=\"mod_types\", y=\"delta_tr\", data=df_vp, order=mod_order, color='k', size=2, alpha=0.6)\n",
    "\n",
    "plt.axhline(0, color='gray', linestyle='--', linewidth=1)\n",
    "plt.xticks(rotation=45)\n",
    "plt.xlabel(\"Modification Type\")\n",
    "plt.ylabel(\"ΔRT (min)\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"img/rt_shift.svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d009e5c-1240-492d-bfcd-8e5000aed2ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"report.tsv\",sep=\"\\t\")\n",
    "df_filtered = df[df[\"Q.Value\"] < 0.01]\n",
    "df_tr = pd.read_csv(\"prosit_ptm_2020.csv\")\n",
    "\n",
    "df_grad_length = abs(df[\"iRT\"].max()-df[\"iRT\"].min())\n",
    "df_grad_length_ptm = abs(df_tr[\"tr\"].max()-df_tr[\"tr\"].min())\n",
    "\n",
    "diann_predictions = df_filtered[\"RT\"] - df_filtered[\"Predicted.RT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1440330d-5660-418f-8519-6927185209bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tr.fillna(\"\",inplace=True)\n",
    "\n",
    "# Separate unmodified and modified peptides\n",
    "df_unmod = df_tr[df_tr[\"modifications\"] == \"\"].rename(columns={\"tr\": \"tr_unmod\"})\n",
    "df_mod = df_tr[df_tr[\"modifications\"] != \"\"]\n",
    "\n",
    "# Merge on sequence to get corresponding unmodified tr for each modified one\n",
    "df_merged = pd.merge(df_mod, df_unmod[[\"seq\", \"tr_unmod\"]], on=\"seq\", how=\"left\")\n",
    "\n",
    "# Calculate RT difference\n",
    "df_merged[\"delta_tr\"] = df_merged[\"tr\"] - df_merged[\"tr_unmod\"]\n",
    "\n",
    "# Optional: sort for readability\n",
    "df_merged = df_merged.sort_values(by=\"seq\")\n",
    "\n",
    "# Step 2: Extract only the modification types from the modification string\n",
    "# Example: \"18|Methyl|9|Carbamidomethyl\" -> [\"Methyl\", \"Carbamidomethyl\"]\n",
    "def extract_mod_types(mod_str):\n",
    "    if not mod_str:\n",
    "        return []\n",
    "    parts = mod_str.split('|')\n",
    "    return [parts[i] for i in range(1, len(parts), 2)]\n",
    "\n",
    "# Create a column of modification lists\n",
    "df_merged[\"mod_types\"] = df_merged[\"modifications\"].apply(extract_mod_types)\n",
    "df_merged = df_merged[df_merged[\"mod_types\"].apply(len) == 1].copy()\n",
    "df_merged[\"mod_types\"] = [mt[0] for mt in df_merged[\"mod_types\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f68d9c86-50c2-433a-9d0f-1adcab411d94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example: Using df_merged with columns \"delta_tr\" and \"mod_types\"\n",
    "values_col = (df_merged[\"delta_tr\"]) / df_grad_length_ptm * 100\n",
    "df_merged[\"adjusted_delta_tr\"] = values_col  # Create a new column for easier plotting\n",
    "\n",
    "selected_mods = [\"Methyl\", \"Propionyl\", \"Dimethyl\", \"Crotonyl\"]\n",
    "# Plot KDE for each mod_type group\n",
    "for mod_type in selected_mods:\n",
    "    subset = df_merged[df_merged[\"mod_types\"] == mod_type]\n",
    "    sns.kdeplot(subset[\"adjusted_delta_tr\"], bw_adjust=0.5, fill=True, alpha=0.1, label=mod_type)\n",
    "\n",
    "sns.kdeplot(diann_predictions, bw_adjust=0.5, color='blue', fill=True, alpha=0.8,  label='Identifications \"unmodified\"')\n",
    "plt.xlim(-10,30)\n",
    "plt.legend()\n",
    "plt.xlabel('Relative retention time difference (%)')\n",
    "plt.ylabel('Density')\n",
    "\n",
    "plt.legend(title='Modification Types')\n",
    "plt.savefig(\"img/retention_time_prediction_real_exp.svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9146c68d-8939-42b4-912f-90123118046b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
